csrc/id_model/id_graph.h

```diff
@@ -0,0 +1,191 @@
+  // ExprGroups used in this history of defining the 'of' IdGroups.
+  ExprGroups allDefinitionsOf(const IdGroups& of) const;
+
+  //! Returns the pointer to expressions associated with the
+  //! definitions of the provided IterDomain group in the provided
+  //! mapping mode (if it exists). Nullptr is returned otherwise.
+  //!
+  //! The returned pointer is to a vector of vector of expressions. The
+  //! inner vector is proven to be equivalent based on the provided mode. The
+  //! outer vector are expression groups that are not equivalent based on the
+  //! provided mode, but produce one of the IterDomains within the same disjoint
+  //! Iter Domain set based on the provided mode.
+  const ExprGroups* getDefinitions(const IdGroup& id_group) const;
+
+  //! Same as iterDomainGroupDefinitions but for uses instead of
+  //! definitions
+  const ExprGroups* getUses(const IdGroup& id_group) const;
+
+  bool hasDefinitions(const IdGroup& id_group) const;
+
+  bool hasUses(const IdGroup& id_group) const;
+
+  std::string toString() const;
+
+  // Checks if the expression is a trivial operation where an input is simply an
+  // output of the transformation. Returns the mapped iter domains if found.
+  static std::vector<std::vector<IterDomain*>> isTrivialExpr(Expr* expr);
+
+  // Returns if all atributes of the ID transforms first and second are the same
+  static bool transformAtributesMatch(Expr* first, Expr* second);
+
+  // Initializes entries for the provided IterDomain in the IterDomainGraphs
+  void initializeId(
+      IterDomain* id,
+      const VectorOfUniqueEntries<Expr*>& definitions,
+      const VectorOfUniqueEntries<Expr*>& uses);
+
+  // Returns if first and second are expressions through which the provided
+  // id_map have matching inputs (if forward), or outputs (if not forward).
+  // Returning true means the expressions are "the same", in terms they modify
+  // matching original extents, by the same amount.
+  bool exprsMap(Expr* first, Expr* second, bool forward) const;
+
+ public:
+  void addUniqueUses(const IdGroup& id_group, const ExprGroup& uses) {
+    unique_uses_.at(id_group).pushBack(uses);
+  }
+
+  void addUniqueDefinitions(const IdGroup& id_group, const ExprGroup& defs) {
+    unique_definitions_.at(id_group).pushBack(defs);
+  }
+
+  // Set id0 and id1 to mapped in disjointIdsSet[mode], attempt to propagate
+  // new mapping through id0/id1 definitions/uses.
+  void mapIds(IterDomain* id0, IterDomain* id1);
+
+  // Checks if expr0 and expr1 should map together, maps them together, and if
+  // expression propagation is on, propagates mapping through them. This should
+  // be the only call in IdGraph to mapThroughExpr
+  void maybeMapThroughExprs(Expr* expr0, Expr* expr1, bool forward);
+
+  // Map through loop swizzles, as input/output IterDomains are exact, only the
+  // order they're traversed differs.
+  void mapThroughLoopSwizzles();
```

Comment:
Could you leave a comment that we need to redo swizzle, and need to revisit if we really want to map through it in the exact map?

Reply:
Added to `buildExactGraph` in `IdModel`.